internal/postgres/reconcile.go

```diff
@@ -258,3 +261,146 @@ func PodSecurityContext(cluster *v1beta1.PostgresCluster) *corev1.PodSecurityCon
 
 	return podSecurityContext
 }
+
+// GenerateUpgradeJobIntent creates an pg_upgrade Job to perform a major
+// PostgreSQL upgrade.
+func GenerateUpgradeJobIntent(
+	cluster *v1beta1.PostgresCluster, sa string,
+	spec *v1beta1.PostgresInstanceSetSpec,
+	inClusterCertificates, inClientCertificates *corev1.SecretProjection,
+	inDataVolume, inWALVolume *corev1.PersistentVolumeClaim,
+) (batchv1.Job, error) {
+
+	// create the pg_upgrade Job
+	upgradeJob := &batchv1.Job{}
+	upgradeJob.ObjectMeta = naming.PGUpgradeJob(cluster)
+
+	// set labels and annotations
+	var labels, annotations map[string]string
+	labels = naming.Merge(cluster.Spec.Metadata.GetLabelsOrNil(),
+		cluster.Spec.Upgrade.Metadata.GetLabelsOrNil(),
+		naming.PGUpgradeJobLabels(cluster.Name))
+	annotations = naming.Merge(cluster.Spec.Metadata.GetAnnotationsOrNil(),
+		cluster.Spec.Upgrade.Metadata.GetAnnotationsOrNil(),
+	)
+	upgradeJob.ObjectMeta.Labels = labels
+	upgradeJob.ObjectMeta.Annotations = annotations
+
+	certVolumeMount := corev1.VolumeMount{
+		Name:      naming.CertVolume,
+		MountPath: naming.CertMountPath,
+		ReadOnly:  true,
+	}
+	certVolume := corev1.Volume{
+		Name: certVolumeMount.Name,
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				// PostgreSQL expects client certificate keys to not be readable
+				// by any other user.
+				// - https://www.postgresql.org/docs/current/libpq-ssl.html
+				DefaultMode: initialize.Int32(0o600),
+				Sources: []corev1.VolumeProjection{
+					{Secret: inClusterCertificates},
+					{Secret: inClientCertificates},
+				},
+			},
+		},
+	}
+
+	dataVolumeMount := DataVolumeMount()
+	dataVolume := corev1.Volume{
+		Name: dataVolumeMount.Name,
+		VolumeSource: corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: inDataVolume.Name,
+				ReadOnly:  false,
+			},
+		},
+	}
+	//while true; do sleep 30; done;
+	script := fmt.Sprintf(`OLDVERSION=%d
```

Comment:
:pick:  It'd be good to move this script into its own function (e.g. such as `startupCommand()` or `reloadCommand()`) so that we can run the script through a shellcheck test (such as we do in `TestStartupCommand()`, etc.).

:wrench: In anticipation of an eventual ShellCheck test, we should align with their guidance.  Therefore, the bash vars should be defined using `"${NEWVERSION}"` rather than `$NEWVERSION`.

Reply:
:+1: I moved the script to a new function and added a shellcheck test.